app/src/ui/rebase/warn-force-push-dialog.tsx

```diff
@@ -0,0 +1,104 @@
+import * as React from 'react'
+
+import { Repository } from '../../models/repository'
+import { WarnForcePushStep } from '../../models/rebase-flow-step'
+
+import { Button } from '../lib/button'
+import { ButtonGroup } from '../lib/button-group'
+import { Checkbox, CheckboxValue } from '../lib/checkbox'
+
+import { Dispatcher } from '../dispatcher'
+import { DialogFooter, DialogContent, Dialog } from '../dialog'
+
+interface IWarnForcePushProps {
+  readonly dispatcher: Dispatcher
+  readonly repository: Repository
+  readonly step: WarnForcePushStep
+  readonly askForConfirmationOnForcePush: boolean
+  readonly onDismissed: () => void
+}
+
+interface IWarnForcePushState {
+  readonly askForConfirmationOnForcePush: boolean
+}
+
+export class WarnForcePushDialog extends React.Component<
+  IWarnForcePushProps,
+  IWarnForcePushState
+> {
+  public constructor(props: IWarnForcePushProps) {
+    super(props)
+
+    this.state = {
+      askForConfirmationOnForcePush: props.askForConfirmationOnForcePush,
+    }
+  }
+
+  public render() {
+    const { baseBranch, targetBranch } = this.props.step
+
+    return (
+      <Dialog
+        title="This rebase will require a force push"
+        onDismissed={this.props.onDismissed}
+        onSubmit={this.onContinueRebase}
+        disableClickDismissalAlways={true}
+        type="warning"
+      >
+        <DialogContent>
+          <p>
+            Are you sure you want to rebase <strong>{targetBranch.name}</strong>{' '}
+            onto <strong>{baseBranch.name}</strong>?
+          </p>
+          <p>
+            At the end of the rebase flow, GitHub Desktop will enable you to
+            force push the branch to update the upstream branch. Force pushing
+            will alter the history on the remote and potentially cause problems
+            for others collaborating on this branch.
+          </p>
+          <div>
+            <Checkbox
+              label="Do not show this message again"
+              value={
+                this.state.askForConfirmationOnForcePush
+                  ? CheckboxValue.Off
+                  : CheckboxValue.On
+              }
+              onChange={this.onAskForConfirmationOnForcePushChanged}
+            />
+          </div>
+        </DialogContent>
+        <DialogFooter>
+          <ButtonGroup>
+            <Button type="submit">Continue rebase</Button>
```

Comment:
Suggestion per our discussion in the thread.
```suggestion
            <Button type="submit">Begin rebase</Button>
```